File: endpoints/photos.py
from silence.decorators import endpoint

@endpoint(
    route="/photos",
    method="GET",
    sql="SELECT * FROM Photos"
)
def get_all():
    pass

###############################################################################

@endpoint(
    route="/photos/$photoId",
    method="GET",
    sql="SELECT photos.*,username,profilePicUrl,AVG(score)AS score,GROUP_CONCAT(DISTINCT categoryName)AS categories FROM photos LEFT JOIN categoriesphotos ON photos.photoId=categoriesphotos.photoId LEFT JOIN categories ON categoriesphotos.categoryId=categories.categoryId LEFT JOIN ratings ON photos.photoId=ratings.photoId JOIN users ON photos.userId=users.userId WHERE photos.photoId = $photoId  GROUP BY categoriesphotos.photoId"
)
def get_by_id():
    pass

###############################################################################

@endpoint(
    route="/photos/recent",
    method="GET",
    sql="SELECT * FROM photos WHERE visibility='PUBLIC' ORDER BY uploadDate DESC"
)
def get_recent():
    pass

###############################################################################

@endpoint(
    route="/photos/recent/$userId",
    method="GET",
    sql="SELECT * FROM photos WHERE visibility='PRIVATE' AND userId=$userId ORDER BY uploadDate DESC"
)
def get_recent_private():
    pass

###############################################################################

@endpoint(
    route="/photos/users/$userId",
    method="GET",
    sql="SELECT photos.*,users.username FROM Users NATURAL JOIN Photos WHERE userId = $userId AND visibility='PUBLIC' ORDER BY uploadDate DESC"
)
def get_user_posts():
    pass

###############################################################################
@endpoint(
    route="/photos/categories/$categoryId",
    method="GET",
    sql="SELECT photos.*,categories.categoryName FROM photos NATURAL JOIN categoriesphotos NATURAL JOIN categories WHERE categoryId = $categoryId AND visibility='PUBLIC' ORDER BY uploadDate DESC"
)
def get_category_posts():
    pass

###############################################################################

@endpoint(
    route="/photos/top_rated",
    method="GET",
    sql="SELECT photos.* FROM ratings JOIN photos ON ratings.photoId=photos.photoId AND (70>(DATEDIFF(CURRENT_TIMESTAMP, uploadDate))) WHERE visibility='PUBLIC' GROUP BY photos.photoId ORDER BY AVG(score) DESC LIMIT 5"
)
def get_top_rated():
    pass

###############################################################################

@endpoint(
    route="/photos/top_commented",
    method="GET",
    sql="SELECT photos.* FROM comments JOIN photos ON comments.photoId=photos.photoId AND (70>(DATEDIFF(CURRENT_TIMESTAMP, uploadDate))) WHERE visibility='PUBLIC' GROUP BY photos.photoId ORDER BY COUNT(commentId) DESC LIMIT 5"
)
def get_top_commented():
    pass

###############################################################################

@endpoint(
    route="/photos/followedBy/$userId",
    method="GET",
    sql="SELECT photos.* FROM photos NATURAL JOIN users JOIN usersusers ON users.userId=usersusers.userFollowsId WHERE userFollowedById=$userId AND visibility='PUBLIC' ORDER BY uploadDate DESC "
)
def get_followed():
    pass

###############################################################################

@endpoint(
    route="/photos",
    method="POST",
    sql="INSERT INTO Photos (title, description, photoUrl, visibility, userId) VALUES ($title, $description, $photoUrl, $visibility, $userId)",
    description="Creates a new photo",
    auth_required=False,
)
def create(title, description, photoUrl, visibility, userId):
    pass


###############################################################################

@endpoint(
    route="/photos/$photoId",
    method="PUT",
    sql="UPDATE Photos SET title = $title, description = $description, photoUrl = $photoUrl, visibility = $visibility WHERE photoId = $photoId",
    description="Updates an existing photo",
    auth_required=False,
)
def update(title, description, photoUrl, visibility):
    pass

###############################################################################

@endpoint(
    route="/photos/$photoId",
    method="DELETE",
    sql="DELETE FROM Photos WHERE photoId = $photoId",
    description="Removes a photo",
    auth_required=False,
)
def delete():
    pass

File: docs/js/api/photos.js
"use strict";
import { BASE_URL, requestOptions } from "/js/api/common.js";

const photosAPI = {
    getAll: function(){
        return new Promise(function(resolve,reject){
            axios.get(BASE_URL + "/photos")
                .then(response => resolve(response.data))
                .catch(error => reject(error.response.data.message));
        });
    },
    getRecent: function(){
        return new Promise(function(resolve,reject){
            axios.get(BASE_URL + "/photos/recent")
                .then(response => resolve(response.data))
                .catch(error => reject(error.response.data.message));
        });
    },
    getRecentPrivate: function(userId){
        return new Promise(function(resolve,reject){
            axios.get(BASE_URL + "/photos/recent/"+userId)
                .then(response => resolve(response.data))
                .catch(error => reject(error.response.data.message));
        });
    },

    getById: function(photoId){
        return new Promise(function(resolve,reject){
            axios.get(BASE_URL + "/photos/" + photoId)
                .then(response => resolve(response.data[0]))
                .catch(error => reject(error.response.data.message));
        });
    },
    getTopRated: function(){
        return new Promise(function(resolve,reject){
            axios.get(BASE_URL + "/photos/top_rated")
                .then(response => resolve(response.data))
                .catch(error => reject(error.response.data.message));
        });
    },
    getTopCommented: function(){
        return new Promise(function(resolve,reject){
            axios.get(BASE_URL + "/photos/top_commented")
                .then(response => resolve(response.data))
                .catch(error => reject(error.response.data.message));
        });
    },
    getUserPosts: function(userId){
        return new Promise(function(resolve,reject){
            axios.get(BASE_URL + "/photos/users/" + userId)
                .then(response => resolve(response.data))
                .catch(error => reject(error.response.data.message));
        });
    },
    getCategoryPosts: function(categoryId){
        return new Promise(function(resolve,reject){
            axios.get(BASE_URL + "/photos/categories/" + categoryId)
                .then(response => resolve(response.data))
                .catch(error => reject(error.response.data.message));
        });
    },
    getFollowedPosts: function(userId){
        return new Promise(function(resolve,reject){
            axios.get(BASE_URL + "/photos/followedBy/" + userId)
                .then(response => resolve(response.data))
                .catch(error => reject(error.response.data.message));
        });
    },

    create: function (formData) {
        return new Promise (function(resolve,reject){
            axios.post(BASE_URL + "/photos", formData)
                .then(response => {
                    resolve (response.data);
                })
                .catch(error => reject (error.response.data.message));
        });
    },
    update: function (photoId,formData) {
        return new Promise (function(resolve,reject){
            axios.put(BASE_URL + "/photos/" + photoId , formData)
                .then(response => resolve (response.data))
                .catch(error => reject (error.response.data.message));
        });
    },
    delete: function (photoId) {
        return new Promise (function(resolve,reject){
            axios.delete(BASE_URL + "/photos/" + photoId)
                .then(response => resolve (response.data))
                .catch(error => reject (error.response.data.message));
        });
    },
    
        
    
};


export { photosAPI };